src/common/bfloat16.hpp

```diff
@@ -31,7 +31,11 @@ namespace impl {
 struct bfloat16_t {
     uint16_t raw_bits_;
     bfloat16_t() = default;
-    constexpr bfloat16_t(uint16_t r, bool) : raw_bits_(r) {}
+    constexpr bfloat16_t(uint16_t r, bool) :
```

Comment:
(Random spot) JFYI: the clang-format check is failing.

Reply:
@rsdubtso  Thanks very much for your useful feedback. Do you mean that the `bfloat16_t` implementation should _only_ treat denormals as zero (or flush them to zero) during `float` to `bfloat16_t` conversion and as part of the implementation of the dot product ?

> Despite this being 'reference' unoptimized code path, I would be wary of adding any extra checks without looking into the performance impact first.

Another approach could be to just have a debug-only assertion that the `r` argument of `bfloat16_t(uint16_t r, bool)` does not have the bit pattern of a SNaN or a denormal:

    assert( ! is_signaling_NaN(r) && ! is_denormal(r) );

What do you think? 

> That said, I think it would be nice to understand what causes the failures in [tensorflow/tensorflow#36514](https://github.com/tensorflow/tensorflow/issues/36514).

Surely I'm interested to hear their opinion as well on denormals and bfloat16  😃

> JFYI: the clang-format check is failing.

Pleas let me know if there is any way to see more detailed output from azure on this clang-format failure than just "[Clang-format check FAILED! Found not formatted files!](https://dev.azure.com/mkldnn/241bf441-2b73-415e-aad0-6b0e5e04fcfe/_apis/build/builds/69/logs/27)"

Reply:
> Another approach could be to just have a debug-only assertion that the r argument of bfloat16_t(uint16_t r, bool) does not have the bit pattern of a SNaN or a denormal:

I would not be very much against this. I just want to understand what such check would enforce because SNaNs and denormals are benign during normal bfloat16 operations. Probably I am missing something.

> Please let me know if there is any way to see more detailed output from azure on this clang-format failure than just "Clang-format check FAILED! Found not formatted files!"

I don't know of any such way. What usually helps is either having a clang-format git hook, or you can just run run `clang-format -i -style=file` on modified files manually.